src/emmy/expression/compile.cljc

```diff
@@ -499,31 +300,51 @@
   [f-form]
   (sci/eval-form (sci/fork sci-context) f-form))
 
-(defn- compile-state->source
-  "Returns a natively-evaluated Clojure function that implements `body`, given:
+(defn- compile->clj
+  "Returns Clojure source for a function that implements `body`, given:
 
   - `params`: a seq of symbols equal in count to the original state function's
     args
   - `state-model`: a sequence of variables representing the structure of the
     nested function returned by the state function
   - `body`: a function body making use of any symbol in the args above"
-  [params state-model body opts]
+  [argv body]
   (let [body (w/postwalk-replace sym->resolved-form body)]
-    `(fn ~(state-argv params state-model opts) ~body)))
+    `(fn [~@argv] ~body)))
+
+(defn- compile->js
+  "Returns an array containing JavaScript source for a function that implements
+  `body` in a form suitable for the arguments of the Function constructor, given:
+
+  - `params`: a seq of symbols equal in count to the original state function's
+    args
+  - `state-model`: a sequence of variables representing the structure of the
+    nested function returned by the state function
+  - `body`: a function body making use of any symbol in the args above"
+  [argv body]
+  (let [argv        (mapv commafy-arglist argv)
+        js          (render/->JavaScript* body)
+        cs-bindings (s/join (for [[var val] (:vars js)]
+                              (str "  const " var " = " val ";\n")))
+        body        (str cs-bindings
+                         "  return "
+                         (:value js)
+                         ";")]
+    (conj argv body)))
 
-(defn- compile-state-native
-  "Returns a natively-evaluated Clojure function that implements `body`, given:
+(defn- compile-native
+  "Returns a natively-evaluated function that implements `body`, given:
 
   - `params`: a seq of symbols equal in count to the original state function's
     args
   - `state-model`: a sequence of variables representing the structure of the
     nested function returned by the state function
   - `body`: a function body making use of any symbol in the args above"
-  [params state-model body opts]
-  (eval
-   (compile-state->source params state-model body opts)))
+  [argv body]
+  #?(:clj (eval (compile->clj argv body))
```

Comment:
rather than compose with `js->clj`, what do you think of making an explicit call inside the ->JavaScript` function to a clojure vector constructor that we export? That way we could use this same function to compile functions that mutate the output vector, or return native js arrays by design. Thoughts?

Reply:
I put this in because I was getting a weird error of "IReduce.-reduce" not being present which I tracked down to the JS-nature of the returned array. This caused me to conclude that a JS array looked innocent but could fail to work in some distant context. Supplying a cli vector compatible constructor would do the trick, at the cost of the compiled code not looking like "real" JS. 

But you point out that this is a copy in performance critical code ... I'd rather "fix" this with an `aset` output convention than inject the constructor. I have personally benefited from the ability to paste the output into browser code, so I want to keep it.